lib/core/src/chain_swap.rs

```diff
@@ -0,0 +1,382 @@
+                            if let Some(transaction) = update.transaction.clone() {
+                                self.update_swap_info(
+                                    id,
+                                    Pending,
+                                    Some(&transaction.id),
+                                    None,
+                                    None,
+                                    None,
+                                )
+                                .await?;
+                            }
+                            self.claim(&swap).await.map_err(|e| {
+                                error!("Could not cooperate Chain Swap {id} claim: {e}");
+                                anyhow!("Could not post claim details. Err: {e:?}")
+                            })?;
+                        }
+                        _ => info!("Waiting for server lockup confirmation for Chain Swap {id}"),
+                    },
+                    Some(claim_tx_id) => {
+                        warn!("Claim tx for Chain Swap {id} was already broadcast: txid {claim_tx_id}")
+                    }
+                };
+                Ok(())
+            }
+
+            // If swap state is unrecoverable, either:
+            // 1. The swap has expired (>24h)
+            // 2. The transaction failed
+            // 3. Lockup failed (too little funds were sent)
+            // We initiate a cooperative refund, and then fallback to a regular one
+            ChainSwapStates::TransactionFailed
+            | ChainSwapStates::TransactionLockupFailed
+            | ChainSwapStates::SwapExpired => {
+                match swap.refund_tx_id {
+                    None => {
+                        warn!("Chain Swap {id} is in an unrecoverable state: {swap_state:?}");
+                        match (swap.payment_type, swap.user_lockup_tx_id.clone()) {
+                            // If there is a user lockup tx when sending we need to refund it
+                            (PaymentType::Send, Some(_)) => {
+                                warn!("Chain Swap {id} user lockup tx has been broadcast. Attempting refund.");
+                                let refund_tx_id = self.refund(&swap).await?;
+                                info!("Broadcast refund tx for Chain Swap {id}. Tx id: {refund_tx_id}");
+                                self.update_swap_info(
+                                    id,
+                                    Pending,
+                                    None,
+                                    None,
+                                    None,
+                                    Some(&refund_tx_id),
+                                )
+                                .await?;
+                            }
+
+                            // If there is a lockup tx when receiving we need to refund to a sender address
+                            // TODO: Get the sender refund address
```

Comment:
I see two ways to do this:
- the SDK asks for a refund address each time an onchain swap is started
- the SDK config contains a BTC xpub, which is set on init, from which we can derive new refund addresses when needed

The xpub way is probably best, @roeierez what do you think?

Reply:
If we refund back to L-BTC, then it's simpler: `sdk.next_unused_address()` should do the trick.

Reply:
The refund is on the BTC lockup side, so it should be the sender getting the refund.

> If we refund back to L-BTC, then it's simpler: `sdk.next_unused_address()` should do the trick.

Yes, the L-BTC lockup side refund is covered like this

Reply:
> it should be the sender getting the refund

Then looks like this should do the trick:

```rust
let sender_refund_address = swap.get_boltz_create_response()?.lockup_details.refund_address;
```

Reply:
OK, I see there is another PR for receive so I will ignore that in my review.

Reply:
@dangeross @ok300 I started to review the other PR and I see the intention is to handle refund here so here are my thoughts:
I think refunding to an external BTC address should be manual (user API) and not automatic from the following reasons:
1. It allow the user defines the fee at the time of refund
2. It allows the user bump the fees for the refund tx if needed.
3. It is consistent with the breez-sdk

So what do you think about adding the following endpionts:
1. list_refundables
2. refund

And ignore refundable states in the chain_swap.rs

Reply:
This is only for the receive case where we need the BTC address right? Otherwise we handle it automatically in the send case

Reply:
Yes, exactly.